File: implementations/php/lib/Eval/Tapered.php
<?php

namespace Chess\Eval;

use Chess\Board;

final class Tapered {
    public static function computePhase(Board $board): int {
        $phase = 0;
        for ($row = 0; $row < 8; $row++) {
            for ($col = 0; $col < 8; $col++) {
                [$piece, ] = $board->get_piece($row, $col);
                if ($piece === CHESS_EMPTY) {
                    continue;
                }
                $phase += Tables::phaseValue($piece);
            }
        }

        return min(Tables::totalPhase(), $phase);
    }

    public static function evaluateMaterialAndPst(Board $board): array {
        $middlegame = 0;
        $endgame = 0;

        for ($row = 0; $row < 8; $row++) {
            for ($col = 0; $col < 8; $col++) {
                [$piece, $color] = $board->get_piece($row, $col);
                if ($piece === CHESS_EMPTY) {
                    continue;
                }

                $sign = $color === CHESS_WHITE ? 1 : -1;
                $value = CHESS_PIECE_VALUES[$piece];
                $middlegame += $sign * ($value + Tables::middlegamePst($piece, $row, $col, $color));
                $endgame += $sign * ($value + Tables::endgamePst($piece, $row, $col, $color));
            }
        }

        return [$middlegame, $endgame];
    }

    public static function interpolateScore(int $phase, int $middlegame, int $endgame): int {
        return intdiv(($middlegame * $phase) + ($endgame * (256 - $phase)), 256);
    }

    public static function scalePhaseTo256(int $phase): int {
        return intdiv($phase * 256, Tables::totalPhase());
    }
}

File: implementations/haskell/src/Eval/Tapered.hs
module Eval.Tapered where

import Types
import Board
import Eval.Tables

computePhase :: GameState -> Int
computePhase gs =
  min totalPhase $ sum [phaseValue ptype | sq <- allSquares, Just (Piece _ ptype) <- [getPiece gs sq]]
  where
    allSquares = [Square col row | col <- [0..7], row <- [0..7]]

evaluateMaterialAndPst :: GameState -> (PieceType -> [Int]) -> Int
evaluateMaterialAndPst gs pst =
  sum [pieceScore sq piece | sq <- allSquares, Just piece <- [getPiece gs sq]]
  where
    allSquares = [Square col row | col <- [0..7], row <- [0..7]]
    pieceScore sq (Piece color ptype) =
      let base = pieceBaseValue ptype
          pstValue = squareTableValue (pst ptype) color sq
          total = base + pstValue
      in if color == White then total else -total

squareTableValue :: [Int] -> Color -> Square -> Int
squareTableValue table color (Square col row) =
  let idx = if color == White then row * 8 + col else (7 - row) * 8 + col
  in table !! idx

interpolateScore :: Int -> Int -> Int -> Int
interpolateScore phase mgScore egScore = (mgScore * scaled + egScore * (256 - scaled)) `div` 256
  where
    scaled = (phase * 256) `div` totalPhase
